cpp/CMakeLists.txt

```diff
@@ -382,6 +375,40 @@ enable_testing()
 ############################################################
 
 # ----------------------------------------------------------------------
+# Thirdparty toolchain
+
+set(THIRDPARTY_DIR "${CMAKE_SOURCE_DIR}/thirdparty")
+set(GFLAGS_VERSION "2.1.2")
+set(GTEST_VERSION "1.8.0")
+set(GBENCHMARK_VERSION "1.1.0")
+set(FLATBUFFERS_VERSION "1.6.0")
+set(JEMALLOC_VERSION "4.4.0")
+
+if (NOT "$ENV{ARROW_BUILD_TOOLCHAIN}" STREQUAL "")
+  if ("$ENV{FLATBUFFERS_HOME}" STREQUAL "")
+    set(FLATBUFFERS_HOME "$ENV{ARROW_BUILD_TOOLCHAIN}")
+  else()
+    set(FLATBUFFERS_HOME "$ENV{FLATBUFFERS_HOME}")
+  endif()
+
+  if ("$ENV{BOOST_ROOT}" STREQUAL "")
+    set(ENV{BOOST_ROOT} "$ENV{ARROW_BUILD_TOOLCHAIN}")
+  endif()
+
+  if ("$ENV{RAPIDJSON_HOME}" STREQUAL "")
+    set(RAPIDJSON_HOME "$ENV{ARROW_BUILD_TOOLCHAIN}")
+  else()
+    set(RAPIDJSON_HOME "$ENV{RAPIDJSON_HOME}")
+  endif()
+
+  if ("$ENV{JEMALLOC_HOME}" STREQUAL "")
+    set(JEMALLOC_HOME "$ENV{ARROW_BUILD_TOOLCHAIN}")
+  else()
+    set(JEMALLOC_HOME "$ENV{JEMALLOC_HOME}")
+  endif()
+endif()
```

Comment:
Argh, I realized this does not handle the case where this environment variable is not set. Will revise